server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -414,7 +419,7 @@ private static boolean handleMappingUpdateRequired(
             ).mappingSource();
             final DocumentMapper existingDocumentMapper = mapperService.documentMapper();
             if (existingDocumentMapper != null && mergedSource.equals(existingDocumentMapper.mappingSource())) {
-                context.resetForNoopMappingUpdateRetry(mapperService.mappingVersion());
+                context.resetForNoopMappingUpdateRetry(initialMappingVersion);
```

Comment:
There is a possibility that we get to a noop mapping retry twice on the same document. Quite subtle, but it goes like this:
1. User does an update with a script, that dynamically generates a new field based on data, either fields or values.
2. An added field needs a mapping update.
3. That mapping update was just added by another thread and thus we get to a noop mapping update retry.
4. The retry sees a new version of the doc and adds another field, since data in the doc changed.
5. The mapping update could again have been added by another thread, getting us to a mapping update retry.

I do not think the change fundamentally changes the problem. But taking it prior to reading the mapping seems to increase the window of opportunity for the second round to have the old mapping version, thus failing the check in `resetForNoopMappingUpdateRetry`. I hence suggest to revert this line from this PR (and we can follow-up addressing this).

Reply:
Makes sense, sorry for the back and forth on this just now, had to adjust a random test to account for the extra call to the mapper version :)